### PR TITLE
D3ASIM-2876: Enabled toggling of allow_external_connection via live e…

### PR DESCRIPTION
**Pull request by hannesdiedrich**
…vent in external strategies.

I tried it with PV and Load strategy and the it worked:

Screenshot 2021-01-20 at 16 28 03

### CODE REVIEW

File: src/d3a/models/strategy/external_strategies/__init__.py [hunk 1]
```diff
@@ -86,11 +86,20 @@ class ExternalMixin:
     def __init__(self, *args, **kwargs):
         self._connected = False
         self.connected = False
+        self.use_template_strategy = False
```
Comment (spyrostz) [inline src/d3a/models/strategy/external_strategies/__init__.py:89]: Would be better to make this parameter private, only in order to discourage other people from accessing it from outside of this class.

**Review by spyrostz — APPROVED** [review]
One minor comment, LGTM otherwise

**Review by faizan2590 — APPROVED** [review]
LGTM

**Comment by hannesdiedrich** [comment]
@spyrostz @faizan2590 In order to enable switching the API ON for a device that uses the template strategy at startup, i would have to move the self._use_template_strategy flag and its assignment into the load_hours, pv, storage strategy. 
I am not a big fan of that and wanted to have your opinion, before I ask Sarah whether this is really needed.

**Comment by spyrostz** [comment]
> @spyrostz @faizan2590 In order to enable switching the API ON for a device that uses the template strategy at startup, i would have to move the self._use_template_strategy flag and its assignment into the load_hours, pv, storage strategy.
> I am not a big fan of that and wanted to have your opinion, before I ask Sarah whether this is really needed.

I would also not like that at all, and I frankly would prefer to deal with allowing live data + template strategies (if it is required of course) on its own story and not in this PR.

File: src/d3a/models/strategy/external_strategies/__init__.py [hunk 2]
```diff
@@ -104,7 +113,8 @@ def is_aggregator_controlled(self):
 
     @property
     def should_use_default_strategy(self):
-        return not self.connected and not self.is_aggregator_controlled
+        return self._use_template_strategy or \
+               (not self.connected and not self.is_aggregator_controlled)
```
Comment (Hassan754) [inline src/d3a/models/strategy/external_strategies/__init__.py:117]:
```suggestion
        return self._use_template_strategy or \
               not (self.connected or self.is_aggregator_controlled)
```
Minor

**Review by Hassan754 — APPROVED** [review]
LGTM , but i could not follow the trace of toggling ( where is the endpoint or redis channel and doesn't it need to be configured from the `d3a-web` side ?

**Comment by hannesdiedrich** [comment]
> LGTM , but i could not follow the trace of toggling ( where is the endpoint or redis channel and doesn't it need to be configured from the `d3a-web` side ?

@Hassan754 The toggling is done by the `allow_external_connection` flag. Have look into here: https://github.com/gridsingularity/d3a/pull/1026/files#diff-bc1fcffcd59eb05e5d8f357d9c2e6d095de871c1ab57e983392af8e51f1ac54eR136

**Review by Hassan754 — APPROVED** [review]

**Review by spyrostz — APPROVED** [review]
LGTM

**Pull request merged**